xcmedid.1
.TH xcmedid 1 "September 24, 2011" "User Commands"
.SH NAME
xcmddc \- monitor EDID to key/value dumper
.SH SYNOPSIS
\fBxcmedid\fR
.fi 
\fBxcmedid\fR \fIEDID_FILE\fR
.fi 
.SH DESCRIPTION
The xcmedid tool converts a edid data block into key/value pairs. The understood keys depend on what is supported by libXcm.
.SH EXAMPLES 
.TP
Show key/value pairs from a stored EDID data block:
.B xcmedid edid.bin
.PP 
.TP
Fetch EDID over i2c and dump key/value pairs:
.B xcmddc -d 0 | xcmedid
.PP 
.TP
Fetch EDID over Oyranos and dump key/value pairs:
.B oyranos-monitor -f edid -d 0 | xcmedid
.PP 
.TP
Fetch EDID over Oyranos and show in ppmcie:
.B ppmcie `oyranos-monitor -f edid -d 0 | xcmedid --ppmcie` > tmp.ppm; xli tmp.ppm
.PP 
.SH AUTHOR
Kai-Uwe Behrmann (ku.b (at) gmx.de)
.SH COPYRIGHT
(c) 2010, Kai-Uwe Behrmann
.fi
License: MIT <http://www.opensource.org/licenses/mit-license.php>
.SH "SEE ALSO"
xcmddc(1) xcmevents(1) libXcm(3) XcmEdidParse(3)
.fi
http://www.oyranos.org
.SH "REPORTING BUGS"
at: http://sourceforge.net/tracker/?group_id=177017&atid=879553


ChangeLog
Version xcm-0.4.2

Kai-Uwe Behrmann (6):
	* [docu]: man page minor fixes
	* [build]: switch to bz2 compression for RPM
	* [docu]: add xcmevents man page
	* [core]: add JSON from EDID output
	* [core]: we support DDC only on Linux
	* [conf]: bump version

Version xcm-0.4.1

Kai-Uwe Behrmann (10):
	* [build]: rpm group is System/Libraries
	* [build]: disable ddc on apple
	* [build]: add XCM_H flags
	* [build]: compress package with bzip2
	* [conf]: sync test script with Oyranos
	* [conf]: use -fPIC only on 64bit builds
	* [conf]: bump version
	* [conf]: use SUBSYTEM="i2c"
	* [build]: add udev to build requirements
	* [docu]: update ChangeLog

Nicolas Chauvet (2):
	* [conf]: apply Fedora patch for pkg-config
	* [build]: ensure the target directory for udev

Version xcm-0.4.0

Kai-Uwe Behrmann (14):
	* [docu]: update README
	* [core]: add xcmddc --identify option
	* [conf]: install and package udev rules
	* [build]: install and package man pages
	* [docu]: add xcmddc and xcmedid man pages
	* [core]: sync help texts of xcmedid and xcmddc
	* [core]: add --i2c and --force-output to xcmddc
	* [docu]: fix source file header
	* [build]: compile xcmddc
	* [core]; add initial xcmddc tool
	* [conf]: fix version macro names
	* [core]: support stdin stream
	* [build]: fixes for Mandriva RPMs
	* [conf]: bump version

Xcm-0.3.0




src/xcmedid/xcmedid.c
/** xcmedid.c
 *
 *  A EEDID to key/value pair convertor.
 *
 *  @par License: 
 *             MIT <http://www.opensource.org/licenses/mit-license.php>
 *  @par Copyright:
 *             (c) 2009-2011 - Kai-Uwe Behrmann <ku.b@gmx.de>
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <X11/Xcm/XcmEdidParse.h>

#include "xcm_version.h"
#include "xcm_macros.h"

void printfHelp(int argc, char ** argv)
{
  fprintf( stderr, "\n");
  fprintf( stderr, "%s %s\n",   argv[0],
                                _("is a EDID parsing tool"));
  fprintf( stderr, "  Xcm v%s config: %s devel period: %s\n",
                  XCM_VERSION_NAME,
                  XCM_CONFIG_DATE, XCM_DATE );
  fprintf( stderr, "\n");
  fprintf( stderr, "%s\n",                 _("Usage"));
  fprintf( stderr, "      %s EDID.bin\n\n", argv[0]);
  fprintf( stderr, "\n");


  fprintf( stderr, "%s\n",                 _("Usage"));
  fprintf( stderr, "  %s\n",               _("Print EDID values:"));
  fprintf( stderr, "      %s EDID.bin\n\n", argv[0]);
  fprintf( stderr, "      cat EDID.bin | %s\n", argv[0]);
  fprintf( stderr, "        [--openicc|--ppmcie]\n");
  fprintf( stderr, "\n");
  fprintf( stderr, "  %s\n",               _("Print a help text:"));
  fprintf( stderr, "      %s -h\n",        argv[0]);
  fprintf( stderr, "\n");
  fprintf( stderr, "  %s\n",               _("General options:"));
  fprintf( stderr, "        --openicc       %s\n", _("use JSON"));
  fprintf( stderr, "        --ppmcie        %s\n", _("ppcmcie compatible"));
  fprintf( stderr, "        -v              %s\n", _("verbose"));
  fprintf( stderr, "\n");
  fprintf( stderr, "\n");
}

int main(int argc, char ** argv)
{
  XCM_EDID_ERROR_e err = XCM_EDID_WRONG_SIGNATURE;
  char * txt = 0;
  FILE * fp = 0;
  char * mem = 0;
  size_t size = 0;
  int s = 0;
  int min_args = 1;
  int print_openicc_json = 0,
      print_ppmcie = 0;
  const char * file_name = NULL;

#ifdef USE_GETTEXT
  setlocale(LC_ALL,"");
#endif

  if(argc > 1)
  {
    int pos = 1, i;
    char *wrong_arg = 0;
    while(pos < argc)
    {
      switch(argv[pos][0])
      {
        case '-':
            for(i = 1; pos < argc && i < strlen(argv[pos]); ++i)
            switch (argv[pos][i])
            {
              case 'v': verbose += 1; break;
              case '-':
                        if(i == 1)
                        {
                             if(OY_IS_ARG("openicc"))
                        { print_openicc_json = 1; i=100; break; }
                        else if(OY_IS_ARG("ppmcie"))
                        { print_ppmcie = 1; i=100; break; }
                        }
                        printfHelp(argc, argv);
                        exit (0);
                        break;
              case '?':
              case 'h':
              default:
                        wrong_arg = argv[pos];
                        break;
            }
            break;
        default:
                        if(!file_name)
                        file_name = argv[pos];
      }
      if( wrong_arg )
      {
       fprintf(stderr, "%s %s\n", _("wrong argument to option:"), wrong_arg);
       printfHelp(argc, argv);
       exit(1);
      }
      ++pos;
    }

  } 

  if(file_name)
  {
    fp = fopen(file_name,"rb");
    if(fp)
    {
      fseek(fp,0L,SEEK_END); 
      size = ftell (fp);
      rewind(fp);
      mem = malloc(size);
      s = fread(mem, sizeof(char), size, fp);
    } else
    {
      fprintf(stderr, "Error: Could not open file - \"%s\"\n", argv[min_args]);
    }
  } else
  {
    int c;

    mem = malloc(65535);
    fp = stdin;
    while(((c = getc(stdin)) != EOF) &&
          size < 65535)
      mem[size++] = c;
  }

  if(verbose)
  {
    fprintf(stderr, "JSON=%d\n", print_openicc_json);
  }

  if(mem && size)
  {
    if(print_openicc_json)
      err = XcmEdidPrintOpenIccJSON( mem, &txt, malloc );
    else if(print_ppmcie)
    {
      XcmEdidKeyValue_s * list = 0;
      int i = 0, count = 0;
      err = XcmEdidParse( mem, &list, &count );
      if(list && count)
      {
        fprintf(stderr, "ppmcie:");
        for(i = 0; i < count; ++i)
        {
          if(strcmp(list[i].key,XCM_EDID_KEY_REDx) == 0)
          { fprintf(stdout," -red %g", list[i].value.dbl); fflush(stdout); }
          else if(strcmp(list[i].key,XCM_EDID_KEY_REDy) == 0)
          { fprintf(stdout," %g", list[i].value.dbl); fflush(stdout); }
          else if(strcmp(list[i].key,XCM_EDID_KEY_GREENx) == 0)
          { fprintf(stdout," -green %g", list[i].value.dbl); fflush(stdout); }
          else if(strcmp(list[i].key,XCM_EDID_KEY_GREENy) == 0)
          { fprintf(stdout," %g", list[i].value.dbl); fflush(stdout); }
          else if(strcmp(list[i].key,XCM_EDID_KEY_BLUEx) == 0)
          { fprintf(stdout," -blue %g", list[i].value.dbl); fflush(stdout); }
          else if(strcmp(list[i].key,XCM_EDID_KEY_BLUEy) == 0)
          { fprintf(stdout," %g", list[i].value.dbl); fflush(stdout); }
          else if(strcmp(list[i].key,XCM_EDID_KEY_WHITEx) == 0)
          { fprintf(stdout," -white %g", list[i].value.dbl); fflush(stdout); }
          else if(strcmp(list[i].key,XCM_EDID_KEY_WHITEy) == 0)
          { fprintf(stdout," %g", list[i].value.dbl); fflush(stdout); }
        }
      }
    }
    else
      err = XcmEdidPrintString( mem, &txt, malloc );

      if(err)
        fprintf( stderr, "Error: %s (%d)\n",
                 XcmEdidErrorToString(err), (int)size);
      else if(txt)
      {
        fprintf(stdout, "%s\n", txt);
        free(txt);
      }
  }

  if(fp && fp != stdin)
    fclose (fp);

  return err;
}
